pkg/minibroker/minibroker.go

```diff
@@ -506,77 +513,170 @@ func (c *Client) connectTiller() (*tiller.Client, func(), error) {
 	return tc, close, nil
 }
 
-func (c *Client) Bind(instanceID, serviceID string, bindParams map[string]interface{}) (map[string]interface{}, error) {
-	config, err := c.coreClient.CoreV1().ConfigMaps(c.namespace).Get(instanceID, metav1.GetOptions{})
+// Bind the given service instance (of the given service) asynchronously; the
+// binding operation key is returned.
+func (c *Client) Bind(instanceID, serviceID, bindingID string, acceptsIncomplete bool, bindParams map[string]interface{}) (string, error) {
+	config, err := c.getConfigMap(instanceID)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			msg := fmt.Sprintf("could not find configmap %s/%s", c.namespace, instanceID)
-			return nil, osb.HTTPStatusCodeError{
+			return "", osb.HTTPStatusCodeError{
 				StatusCode:   http.StatusNotFound,
 				ErrorMessage: &msg,
 			}
 		}
-		return nil, err
+		return "", err
 	}
 	releaseNamespace := config.Data[ReleaseNamespaceKey]
 	rawProvisionParams := config.Data[ProvisionParamsKey]
+	operationName := generateOperationName(OperationPrefixBind)
 
 	var provisionParams map[string]interface{}
 	err = json.Unmarshal([]byte(rawProvisionParams), &provisionParams)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not unmarshall provision parameters for instance %q", instanceID)
+		return "", errors.Wrapf(err, "could not unmarshall provision parameters for instance %q", instanceID)
 	}
 
-	// Smoosh all the params together
-	params := make(map[string]interface{}, len(bindParams)+len(provisionParams))
-	for k, v := range provisionParams {
-		params[k] = v
-	}
-	for k, v := range bindParams {
-		params[k] = v
+	if acceptsIncomplete {
+		go func() {
+			c.bindSynchronously(instanceID, serviceID, bindingID, releaseNamespace, bindParams, provisionParams)
+		}()
+		return operationName, nil
 	}
 
-	filterByInstance := metav1.ListOptions{
-		LabelSelector: labels.SelectorFromSet(map[string]string{
-			InstanceLabel: instanceID,
-		}).String(),
-	}
+	c.bindSynchronously(instanceID, serviceID, bindingID, releaseNamespace, bindParams, provisionParams)
+	return "", nil
+}
 
-	services, err := c.coreClient.CoreV1().Services(releaseNamespace).List(filterByInstance)
+// bindSynchronously creates a new binding for the given service instance.  All
+// results are only reported via the service instance configmap (under the
+// appropriate key for the binding) for lookup by LastBindingOperationState().
+func (c *Client) bindSynchronously(instanceID, serviceID, bindingID, releaseNamespace string, bindParams, provisionParams map[string]interface{}) {
```

Comment:
This method can fail. It worth returning the error, and handling it properly on line 547.

Reply:
I've updated it do to so (for the synchronous case); I've preferred returning the original error instead of marshalling issues.  Changed in 05d340faf1eeff23091b8d53be40f7bb72104011